### ISSUE TITLE
Screen Reader Usability

### ISSUE THREAD

**Issue by rachel-cmc** [alexed1/LightningFlowComponents#843]
The datatable component is great and I've found it very useful on multiple occasions. However, I have a few screen reader users who have informed me that the component is not accessible to them. They are unable to interact with the component, impeding their ability to work like our other users. We've had to mesh together other workarounds that are less than ideal. 

Can you please consider adding accessibility functionality to your datatable element? Our screen readers can interact with the regular lightning datatables, but unfortunately not your flow datatable.

**Comment by ericrsmith35** [comment]
@rachel-cmc Is it just the Table Header that needs to be identifiable or are there other areas that need to be addressed?

**Comment by rachel-cmc** [comment]
@ericrsmith35 Thanks for getting back to me. The datatable appears to not be coded for accessbility (I'm not a coder, so I cannot verify). On the other hand, normal lightning datatables are screen reader accessible (Please [see this thread](https://trailhead.salesforce.com/trailblazer-community/feed/0D54S00000C5CKMSA3) for some additional context).

What my screen reader users experience with your datatable component for screen flows is:

(1) The table does not read as accessible tables should. When a user navigates through the table they have no idea which column or row they are viewing in the table because it does not announce this to them. The code should declare the column level (row) and column header titles for each spot being viewed. This is what allows the user to know where they are in the table. For example, if you have a table with 3 columns (Name, **Stage**, Date), and 3 rows (header, 1, **2**), when the user is viewing the value in **column 2 row 3** then it should read something like "row 3, Stage, value".

(2) Additionally, the table does not allow screen readers to navigate with standard table navigation keys.

Hopefully that info helps! Let me know if there is additional info I can provide.
Regards,
R

**Comment by rachel-cmc** [comment]
@ericrsmith35 This link was shared with me for context: [https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html](https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html)

**Comment by ericrsmith35** [comment]
This has been implemented in the latest release of Datatable (v3.3.2)
https://unofficialsf.com/datatable-lightning-web-component-for-flow-screens-2/

**Comment by rachel-cmc** [comment]
Hi Eric, thanks for pushing out the new release. I believe I followed the correct steps to get the new version installed. I had my screen reader user test it and she said the experience has not changed. First I installed the new versions of the base packs and then I followed [these tips and tricks](https://unofficialsf.com/tips-tricks-update-upgrade-an-unmanaged-package/#SelectPackage) for updating the datatable component. Is there perhaps a step I missed in the update to the new version? Can you provide any further suggestions?

**Comment by rachel-cmc** [comment]
I realized my flow didn't save the table header I added.... having the user test again when they can now that I know it is there. I'll update once I hear back.

**Comment by ericrsmith35** [comment]
I tested with NVDA before and after the update and I did notice a big
difference.

Eric

**Comment by rachel-cmc** [comment]
Odd, here's what I did:
Retested, and unfortunately it is still not usable for our screen reader user. They are launching the flow from Actions & Recommendations. Tested in multiple browsers.  They are using Jaws. I believe I got the update correct (flow base packs updated, new version of datatable installed, table header is provided in the flow, apex classes added to profile). I still could have missed something... and I'm not with the screen reader user to see them working with the component.

**Comment by ericrsmith35** [comment]
I think they first have to click on an editable cell in the datatable for
the navigation actions to work.

**Comment by rachel-cmc** [comment]
Here's a message from my screen reader user, who switched to NVDA to give it a try in case it was a JAWS issue -

"It’s hard to describe the differences because the over all functionality has not changed.  I can navigate a little of the table using arrow keys, but not with table reading commands.  Have your developer test with NVDA using the key commands control+alt+ left right up or down arrow keys and he’ll see what I mean.  When I use those keyboard short cuts whether with Jaws or NVDA I have the same results.  Neither screen reader flows through the table correctly and does not read the row and column headers as I move through the table. Hope this is helpful."

Note that I'm not actually enabling inline editing within the datatable. They are able to select a row to perform an action on it.

**Comment by rachel-cmc** [comment]
Checking in - Do you mean to say that the datatable has to be in the screen flow with inline editing available in order for the screen reader to interact with the table as might be expected?

**Comment by ericrsmith35** [comment]
I don't think so.  That's just how I was testing it.

**Comment by rachel-cmc** [comment]
Please let me know if you can think of any information I can provide to help describe what my users are experiencing. 

I really appreciate your time and efforts to make this component accessible. I know it would mean a lot to my users if they can interact with the flows in a way similar to my sighted users.... right now, the only way I'm getting it to feasibly work for my screen reader users is by including a link to the related list. But then the users sometimes get confused why a new browser tab opens in certain circumstances (clicking the link in the flow) but not in others (clicking "view all" in a related list, etc.). Allowing them to stay within the flow would certainly simplify things and improve their experience.

**Comment by rachel-cmc** [comment]
I met with one of my users yesterday so they could walk me through what their experience is with the new datatable version. She confirmed that the table navigation commands are not working with either Jaws or NVDA - both of which use the same key combination of ctrl + alt, then the arrows to move throughout. Additionally, the screen reader is still not able to read the context of the row # and column title so the user has the knowledge to understand what is being reviewed in the table. 

Just wanted to confirm what I have in my sandbox's setup where we are testing to see if there may be something I missed:
* FlowActionsBase Pack v2.32
* FlowScreenComponentsBasePack v2.5.4
* datatable v3.3.2
* the user's profile has been assigned the designated apex classes

* the datatable in the flow does have a Table Header
* the datatable is being used without inline editing. Currently, I'm only allowing one row to be selected at a time if an action is to be performed. I do have some datatables being used for multi-row selection and looped actions, but I haven't tested those yet.
* the flow is running on API versions 52.0.
* screen flow is launched via Actions & Recommendations component

**Comment by ericrsmith35** [comment]
I will try and take another look this weekend.

**Comment by rachel-cmc** [comment]
@ericrsmith35 - I wanted to let you know that the datatable IS accessible. There was miscommunication on my end with the screenreader user and confusion with how to activate the action mode since the "Navigation Mode" appears as simple text rather than something to click. Again, I thank you for your time and efforts.
